File: src/lib/mcp/plugins/vm.ts
/* SPDX-License-Identifier: LGPL-2.1-or-later */
import cockpit from "cockpit";
import { ToolPlugin } from "./base.js";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

export class VmPlugin extends ToolPlugin {
    get name(): string {
        return "vm";
    }

    get description(): string {
        return "Virtual Machine management (libvirt/virsh)";
    }

    async detect(): Promise<boolean> {
        try {
            await cockpit.spawn(["which", "virsh"]);
            return true;
        } catch {
            return false;
        }
    }

    register(server: McpServer): void {
        server.registerTool(
            "vm_list",
            {
                title: "List VMs",
                description: "List all VMs with state (running, shut off)",
                inputSchema: z.object({}),
                _meta: { isLowRisk: true },
            },
            async () => {
                const result = await this.runVirsh(["list", "--all"]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_info",
            {
                title: "VM info",
                description: "Get detailed info about a specific VM",
                inputSchema: z.object({
                    name: z.string().describe("Name of the VM")
                }),
                _meta: { isLowRisk: true },
            },
            async ({ name }) => {
                const result = await this.runVirsh(["dominfo", name]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_start",
            {
                title: "Start VM",
                description: "Start a VM",
                inputSchema: z.object({
                    name: z.string().describe("Name of the VM")
                })
            },
            async ({ name }) => {
                const result = await this.runVirsh(["start", name]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_stop",
            {
                title: "Stop VM",
                description: "Shutdown a VM gracefully",
                inputSchema: z.object({
                    name: z.string().describe("Name of the VM")
                })
            },
            async ({ name }) => {
                const result = await this.runVirsh(["shutdown", name]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_reboot",
            {
                title: "Reboot VM",
                description: "Reboot a VM",
                inputSchema: z.object({
                    name: z.string().describe("Name of the VM")
                })
            },
            async ({ name }) => {
                const result = await this.runVirsh(["reboot", name]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_snapshot_list",
            {
                title: "List VM snapshots",
                description: "List snapshots of a VM",
                inputSchema: z.object({
                    name: z.string().describe("Name of the VM")
                }),
                _meta: { isLowRisk: true },
            },
            async ({ name }) => {
                const result = await this.runVirsh(["snapshot-list", name]);
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );

        server.registerTool(
            "vm_create",
            {
                title: "Create VM",
                description: "Create a new VM using virt-install (requires root)",
                inputSchema: z.object({
                    name: z.string().describe("VM Name"),
                    vcpu: z.number().int()
                            .describe("Number of vCPUs"),
                    ram_mb: z.number().int()
                            .describe("RAM in MB"),
                    disk_path: z.string().describe("Path to disk image or zvol (e.g. /dev/zvol/pool/name)"),
                    iso_path: z.string().describe("Path to installer ISO"),
                    os_variant: z.string().optional()
                            .describe("OS variant (e.g. ubuntu22.04)")
                })
            },
            async (args) => {
                try {
                    await cockpit.spawn(["which", "virt-install"]);
                } catch {
                    return {
                        content: [{ type: "text", text: "Error: virt-install is not installed." }],
                        isError: true
                    };
                }
                const cmd = [
                    "virt-install",
                    "--name", args.name,
                    "--vcpus", args.vcpu.toString(),
                    "--memory", args.ram_mb.toString(),
                    "--disk", `path=${args.disk_path}`,
                    "--cdrom", args.iso_path,
                    "--os-variant", args.os_variant || "generic",
                    "--noautoconsole",
                    "--graphics", "vnc"
                ];
                let result = "";

                try {
                    result = await cockpit.spawn(cmd, { superuser: "require" });
                } catch (e: unknown) {
                    return {
                        content: [{ type: "text", text: `Error creating VM: ${e instanceof Error ? e.message : String(e)}` }],
                        isError: true
                    };
                }
                return {
                    content: [{ type: "text", text: result }]
                };
            }
        );
    }

    private async runVirsh(cmdArgs: string[]): Promise<string> {
        try {
            return await cockpit.spawn(["virsh", ...cmdArgs]);
        } catch (e: unknown) {
            return `Error running virsh: ${e instanceof Error ? e.message : String(e)}`;
        }
    }
}

File: src/mcp-server/plugins/vm.py
import shutil
import subprocess
import logging
from typing import List, Dict, Any
from .base import ToolPlugin

logger = logging.getLogger(__name__)

class VmPlugin(ToolPlugin):
    @property
    def name(self) -> str:
        return "vm"

    @property
    def description(self) -> str:
        return "Virtual Machine management (libvirt/virsh)"

    def detect(self) -> bool:
        """Check for virsh and virt-install"""
        return shutil.which("virsh") is not None

    def get_tools(self) -> List[Dict[str, Any]]:
        return [
            {
                "name": "vm_list",
                "description": "List all VMs with state (running, shut off)",
                "inputSchema": {
                    "type": "object",
                    "properties": {},
                }
            },
            {
                "name": "vm_info",
                "description": "Get detailed info about a specific VM",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string", "description": "Name of the VM"}
                    },
                    "required": ["name"]
                }
            },
            {
                "name": "vm_start",
                "description": "Start a VM",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"}
                    },
                    "required": ["name"]
                }
            },
            {
                "name": "vm_stop",
                "description": "Shutdown a VM gracefully",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"}
                    },
                    "required": ["name"]
                }
            },
            {
                "name": "vm_reboot",
                "description": "Reboot a VM",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"}
                    },
                    "required": ["name"]
                }
            },
             {
                "name": "vm_snapshot_list",
                "description": "List snapshots of a VM",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                         "name": {"type": "string"}
                    },
                    "required": ["name"]
                }
            },
            {
                "name": "vm_create",
                "description": "Create a new VM using virt-install (requires root)",
                "inputSchema": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string", "description": "VM Name"},
                        "vcpu": {"type": "integer", "description": "Number of vCPUs"},
                        "ram_mb": {"type": "integer", "description": "RAM in MB"},
                        "disk_path": {"type": "string", "description": "Path to disk image or zvol (e.g. /dev/zvol/pool/name)"},
                        "iso_path": {"type": "string", "description": "Path to installer ISO"},
                        "os_variant": {"type": "string", "description": "OS variant (e.g. ubuntu22.04), optional"}
                    },
                    "required": ["name", "vcpu", "ram_mb", "disk_path", "iso_path"]
                }
            }
        ]

    def execute(self, tool_name: str, args: Dict[str, Any]) -> Any:
        if tool_name == "vm_list":
            return self._run_virsh(["list", "--all"])
        elif tool_name == "vm_info":
            return self._run_virsh(["dominfo", args["name"]])
        elif tool_name == "vm_start":
            return self._run_virsh(["start", args["name"]])
        elif tool_name == "vm_stop":
            return self._run_virsh(["shutdown", args["name"]])
        elif tool_name == "vm_reboot":
            return self._run_virsh(["reboot", args["name"]])
        elif tool_name == "vm_snapshot_list":
             return self._run_virsh(["snapshot-list", args["name"]])
        elif tool_name == "vm_create":
            return self._create_vm(args)
        else:
            raise ValueError(f"Unknown tool: {tool_name}")

    def _run_virsh(self, cmd_args: List[str]) -> str:
        cmd = ["virsh"] + cmd_args
        try:
            result = subprocess.run(cmd, capture_output=True, text=True, check=True)
            return result.stdout
        except subprocess.CalledProcessError as e:
            return f"Error running virsh: {e.stderr}"

    def _create_vm(self, args: Dict[str, Any]) -> str:
        if not shutil.which("virt-install"):
            return "Error: virt-install is not installed."
            
        cmd = [
            "virt-install",
            "--name", args["name"],
            "--vcpus", str(args["vcpu"]),
            "--memory", str(args["ram_mb"]),
            "--disk", f"path={args['disk_path']}",
            "--cdrom", args["iso_path"],
            "--os-variant", args.get("os_variant", "generic"),
            "--noautoconsole", # Don't try to open console viewer
            "--graphics", "vnc" # Use VNC graphics
        ]
        
        try:
            # this takes time, but it backgrounds almost immediately unless waiting for install?
            # virt-install with --noautoconsole returns quickly after starting the VM setup
            result = subprocess.run(cmd, capture_output=True, text=True, check=True)
            return f"VM creation started successfully:\n{result.stdout}"
        except subprocess.CalledProcessError as e:
            return f"Error creating VM: {e.stderr}"
